Fix crashes rendering temporary layers with spatial index

nyalldawson opened the pull request:
Since the underlying libspatialindex is not thread safe on some platforms (e.g. Windows), we have to deep copy the spatial index data if a QgsSpatialIndex is to be accessed across multiple threads (such as when rendering a memory layer with a spatial index attached).

Fixes https://issues.qgis.org/issues/17705 (at the cost of slow feature iterator creation for memory layers with large spatial indexes on Windows - but we don't have an alternative here)

nyalldawson committed:
Fix deep copy of QgsSpatialIndexData

nyalldawson committed:
Add method to force a detach for QgsSpatialIndex

Since the underlying libspatialindex is not thread safe
on some platforms (e.g. Windows), manual calls to detach()
must be made if a QgsSpatialIndex is to be accessed across
multiple threads.

Note that for platforms on which libspatialindex is thread
safe, calling detach() has no effect and does not force the
deep copy.

nyalldawson committed:
Fix memory layers with spatial index crash when rendering

Fixes #17705

nyalldawson committed:
Include job preparation time in layer rendering timer

This prevents layers which are slow to prepare from rendering
in map preview jobs

nyalldawson commented:
@wonder-sk look ok to you?

nirvn commented:
Great catch.

nyalldawson committed:
Avoid deep copy of spatial index when features are fetched on
main thread

wonder-sk commented:
I am wondering whether it wouldn't be better to make QgsSpatialIndex thread-safe by introducing our own locking:
1. having to copy the whole R-Tree every time we render a memory layer is probably going to be a large performance hit
2. having detaches specific to Windows feels wrong and may bite us in some way in the future
3. the maintainers seem to agree that the locking inside libspatialindex as it is now may get removed - see https://github.com/libspatialindex/libspatialindex/issues/59#issuecomment-75381978 ... there is also an unmerged branch which removes thread safety altogether: https://github.com/libspatialindex/libspatialindex/compare/remove-theads
4. using a mutex inside QgsSpatialIndex should not be a significant performance issue because it is relatively uncommon that a single spatial index would be used from multiple threads at once

nyalldawson commented:
@wonder-sk that works for me, I'll adopt

nyalldawson committed:
Implement manual locks on QgsSpatialIndex

Since libspatialindex is not thread safe on all platforms, and
have expressed desire to remove the thread safety that they DO
have on remaining platforms, it's safer and easier for us
to manually add locks to QgsSpatialIndex and be gauranteed that
this class is thread safe on all platforms and libspatialindex
versions.

Also improve docs for the class.

nyalldawson commented:
@wonder-sk like this?

wonder-sk commented:
That looks perfect - thanks!
